src/Catalyst.Core.Modules.Web3/Controllers/Handlers/EthGetStorageAtHandler.cs

```diff
@@ -21,17 +21,48 @@
 
 #endregion
 
+using System;
+using Catalyst.Abstractions.Kvm.Models;
 using Catalyst.Abstractions.Ledger;
-using Nethermind.Core;
+using Catalyst.Protocol.Deltas;
+using LibP2P;
+using Nethermind.Core.Crypto;
+using Nethermind.Dirichlet.Numerics;
+using Nethermind.Store;
+using Address = Nethermind.Core.Address;
 
 namespace Catalyst.Core.Modules.Web3.Controllers.Handlers
 {
     [EthWeb3RequestHandler("eth", "getStorageAt")]
-    public class EthGetStorageAtHandler : EthWeb3RequestHandler<Address, byte[]>
+    public class EthGetStorageAtHandler : EthWeb3RequestHandler<Address, UInt256, BlockParameter, byte[]>
     {
-        protected override byte[] Handle(Address param1, IWeb3EthApi api)
+        protected override byte[] Handle(Address address, UInt256 index, BlockParameter block, IWeb3EthApi api)
         {
-            throw new System.NotImplementedException();
+            Cid deltaHash = GetDeltaHash(api, block);
+
+            Delta delta = api.GetDelta(deltaHash);
+
+            var stateRoot = new Keccak(delta.StateRoot.ToByteArray());
+            api.StateProvider.StateRoot = stateRoot;
+            return api.StorageProvider.Get(new StorageAddress(address, index));
+        }
+
+        static Cid GetDeltaHash(IWeb3EthApi api, BlockParameter block)
+        {
+            switch (block.Type)
+            {
+                case BlockParameterType.Earliest:
+                    return api.DeltaCache.GenesisHash;
+                case BlockParameterType.Latest:
+                    return api.DeltaResolver.LatestDelta;
+                case BlockParameterType.Pending:
+                    return api.DeltaResolver.LatestDelta;
+                case BlockParameterType.BlockNumber:
+                    var blockNumber = block.BlockNumber.Value;
+                    return api.DeltaResolver.Resolve(blockNumber);
+                default:
+                    throw new ArgumentOutOfRangeException();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use a constructor overloads that allows a more meaningful exception message to be provided.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4585091)